lib/handler/pool/state.ex

```diff
@@ -49,23 +51,57 @@ defmodule Handler.Pool.State do
   """
   @spec start_worker(t(), fun, keyword(), pid()) ::
           {:ok, t(), reference()} | {:reject, exception()}
-  def start_worker(state, fun, opts, pid) do
+  def start_worker(state, fun, opts, from_pid) do
     bytes_requested = Handler.max_heap_bytes(opts)
+    task_id = Keyword.get(opts, :task_id)
 
     with :ok <- check_committed_resources(state, bytes_requested),
-         {:ok, ref} <- kickoff_new_task(state, fun, opts) do
-      new_state = commit_resources(state, ref, bytes_requested, pid)
+         {:ok, ref, task_pid} <- kickoff_new_task(state, fun, opts) do
+      new_state = commit_resources(state, ref, bytes_requested, from_pid, task_pid, task_id)
       {:ok, new_state, ref}
     end
   end
 
+  def kill_worker(%State{pool: %Pool{delegate_to: pool}} = state, task_id)
+      when not is_nil(pool) do
+    with :ok <- Pool.kill(pool, task_id) do
+      {:ok, state}
+    end
+  end
+
+  def kill_worker(_state, _task_id = nil),
+    do: {:reject, "Cannot kill task without providing task_id"}
+
+  def kill_worker(state, task_id) do
+    task_not_found = {:reject, "No task with given task_id in state"}
+
+    Enum.reduce_while(state.workers, task_not_found, fn
+      {ref, {_bytes_commited, _from_pid, task_pid, ^task_id}}, _ ->
+        Process.unlink(task_pid)
+        Process.exit(task_pid, :user_killed)
+
+        exception =
+          Handler.ProcessExit.exception(message: "User killed the process", reason: :user_killed)
+
+        state =
+          state
+          |> send_response(ref, {:error, exception})
+          |> cleanup_commitments(ref)
+
+        {:halt, {:ok, state}}
```

Comment:
It looks like we are matching the first task that has a matching `task_id` but we aren't enforcing that they are unique.